File: ios/kwift/shared/Event.swift
import Foundation
//package com.lightningkite.kwift.shared


//import com.lightningkite.kwift.actuals.AnyObject
//import com.lightningkite.kwift.actuals.escaping
//import com.lightningkite.kwift.actuals.weak

final public class Event<T> {
    
    
    init() {
        self.id = 0
    }
    
    var subscriptions = [Subscription<T>]()
    var id: Int
    
    func addAndRun(value: T, listener: @escaping (T) -> Void) -> Subscription<T> {
        listener(value)
        return add(listener : listener)
    }
    func addAndRun(_ value: T, _ listener: @escaping (T) -> Void) -> Subscription<T> { return addAndRun(value: value, listener: listener) }
    
    func add( listener: @escaping (T) -> Void) -> Subscription<T> {
        var element = Subscription(listener : { (item: T) in
            listener(item);
            return false
        }, identifier : id)
        id += 1
        subscriptions.add(element)
        return element
    }
    
    func addWithRemovalCondition( listener: @escaping (T) -> Bool) -> Subscription<T> {
        var element = Subscription(listener : listener, identifier : id)
        id += 1
        subscriptions.add(element)
        return element
    }
    
    func remove(subscription: Subscription<T>) {
        subscriptions.removeAll { (it) in it.identifier == subscription.identifier }
    }
    func remove(_ subscription: Subscription<T>) { remove(subscription: subscription) }
    
    func invokeAll(value: T) {
        subscriptions.removeAll { (subscription) in
            subscription.listener(value)
        }
    }
    func invokeAll(_ value: T) { invokeAll(value: value) }
    
    
    func addWeak<A : AnyObject>(referenceA: A, listener: @escaping (A, T) -> Void) -> Subscription<T> {
        weak var weakA = referenceA;
        var element = Subscription(
            listener : { (item: T) in
                var a = weakA;
                if let a = a {
                    listener(a, item);
                    return false
                } else {
                    return true
                }
            },
            identifier : id
        )
        id += 1
        subscriptions.add(element)
        return element
    }
    func addWeak<A : AnyObject>(_ referenceA: A, _ listener: @escaping (A, T) -> Void) -> Subscription<T> { return addWeak(referenceA: referenceA, listener: listener) }
    
    func addWeak<A : AnyObject, B : AnyObject>(
        referenceA: A,
        referenceB: B, listener: @escaping (A, B, T) -> Void
    ) -> Subscription<T> {
        weak var weakA = referenceA
        weak var weakB = referenceB
        var element = Subscription(
            listener : { (item: T) in
                var a = weakA;
                var b = weakB;
                if let a = a, let b = b {
                    listener(a, b, item);
                    return false
                } else {
                    return true
                }
            },
            identifier : id
        )
        id += 1
        subscriptions.add(element)
        return element
    }
    func addWeak<A : AnyObject, B : AnyObject>(
        _ referenceA: A,
        _ referenceB: B, _ listener: @escaping (A, B, T) -> Void
    ) -> Subscription<T> { return addWeak(referenceA: referenceA, referenceB: referenceB, listener: listener) }
    
    func addWeak<A : AnyObject, B : AnyObject, C : AnyObject>(
        referenceA: A,
        referenceB: B,
        referenceC: C, listener: @escaping (A, B, C, T) -> Void
    ) -> Subscription<T> {
        weak var weakA = referenceA
        weak var weakB = referenceB
        weak var weakC = referenceC
        var element = Subscription(
            listener : { (item: T) in
                var a = weakA;
                var b = weakB;
                var c = weakC;
                if let a = a, let b = b, let c = c {
                    listener(a, b, c, item);
                    return false
                } else {
                    return true
                }
            },
            identifier : id
        )
        id += 1
        subscriptions.add(element)
        return element
    }
    func addWeak<A : AnyObject, B : AnyObject, C : AnyObject>(
        _ referenceA: A,
        _ referenceB: B,
        _ referenceC: C, _ listener: @escaping (A, B, C, T) -> Void
    ) -> Subscription<T> { return addWeak(referenceA: referenceA, referenceB: referenceB, referenceC: referenceC, listener: listener) }
    
    
    func addAndRunWeak<A : AnyObject>(
        referenceA: A,
        value: T,
        listener: @escaping (A, T) -> Void
    ) -> Subscription<T> {
        listener(referenceA, value)
        weak var weakA = referenceA;
        var element = Subscription(
            listener : { (item: T) in
                var a = weakA;
                if let a = a {
                    listener(a, item);
                    return false
                } else {
                    return true
                }
            },
            identifier : id
        )
        id += 1
        subscriptions.add(element)
        return element
    }
    func addAndRunWeak<A : AnyObject>(
        _ referenceA: A,
        _ value: T,
        _ listener: @escaping (A, T) -> Void
    ) -> Subscription<T> { return addAndRunWeak(referenceA: referenceA, value: value, listener: listener) }
    
    func addAndRunWeak<A : AnyObject, B : AnyObject>(
        referenceA: A,
        referenceB: B,
        value: T,
        listener: @escaping (A, B, T) -> Void
    ) -> Subscription<T> {
        listener(referenceA, referenceB, value)
        weak var weakA = referenceA
        weak var weakB = referenceB
        var element = Subscription(
            listener : { (item: T) in
                var a = weakA;
                var b = weakB;
                if let a = a, let b = b {
                    listener(a, b, item);
                    return false
                } else {
                    return true
                }
            },
            identifier : id
        )
        id += 1
        subscriptions.add(element)
        return element
    }
    func addAndRunWeak<A : AnyObject, B : AnyObject>(
        _ referenceA: A,
        _ referenceB: B,
        _ value: T,
        _ listener: @escaping (A, B, T) -> Void
    ) -> Subscription<T> { return addAndRunWeak(referenceA: referenceA, referenceB: referenceB, value: value, listener: listener) }
    
    func addAndRunWeak<A : AnyObject, B : AnyObject, C : AnyObject>(
        referenceA: A,
        referenceB: B,
        referenceC: C,
        value: T,
        listener: @escaping (A, B, C, T) -> Void
    ) -> Subscription<T> {
        listener(referenceA, referenceB, referenceC, value)
        weak var weakA = referenceA
        weak var weakB = referenceB
        weak var weakC = referenceC
        var element = Subscription(
            listener : { (item: T) in
                var a = weakA;
                var b = weakB;
                var c = weakC;
                if let a = a, let b = b, let c = c {
                    listener(a, b, c, item);
                    return false
                } else {
                    return true
                }
            },
            identifier : id
        )
        id += 1
        subscriptions.add(element)
        return element
    }
    func addAndRunWeak<A : AnyObject, B : AnyObject, C : AnyObject>(
        _ referenceA: A,
        _ referenceB: B,
        _ referenceC: C,
        _ value: T,
        _ listener: @escaping (A, B, C, T) -> Void
    ) -> Subscription<T> { return addAndRunWeak(referenceA: referenceA, referenceB: referenceB, referenceC: referenceC, value: value, listener: listener) }
}

File: android/app/src/main/java/com/lightningkite/kwift/shared/Event.kt
package com.lightningkite.kwift.shared

import com.lightningkite.kwift.actuals.AnyObject
import com.lightningkite.kwift.actuals.escaping
import com.lightningkite.kwift.actuals.weak

class Event<T> {
    val subscriptions = ArrayList<Subscription<T>>()
    var id: Int = 0

    fun addAndRun(value: T, @escaping listener: (T) -> Unit): Subscription<T> {
        listener(value)
        return add(listener = listener)
    }

    fun add(@escaping listener: (T) -> Unit): Subscription<T> {
        val element = Subscription(listener = { item: T ->
            listener(item);
            return@Subscription false
        }, identifier = id)
        id += 1
        subscriptions.add(element)
        return element
    }

    fun addWithRemovalCondition(@escaping listener: (T) -> Boolean): Subscription<T> {
        val element = Subscription(listener = listener, identifier = id)
        id += 1
        subscriptions.add(element)
        return element
    }

    fun remove(subscription: Subscription<T>) {
        subscriptions.removeAll { it -> it.identifier == subscription.identifier }
    }

    fun invokeAll(value: T) {
        subscriptions.removeAll { subscription ->
            subscription.listener(value)
        }
    }


    fun <A : AnyObject> addWeak(referenceA: A, @escaping listener: (A, T) -> Unit): Subscription<T> {
        val weakA by weak(referenceA);
        val element = Subscription(
            listener = { item: T ->
                val a = weakA;
                if (a != null) {
                    listener(a, item);
                    return@Subscription false
                } else {
                    return@Subscription true
                }
            },
            identifier = id
        )
        id += 1
        subscriptions.add(element)
        return element
    }

    fun <A : AnyObject, B : AnyObject> addWeak(
        referenceA: A,
        referenceB: B, @escaping listener: (A, B, T) -> Unit
    ): Subscription<T> {
        val weakA by weak(referenceA)
        val weakB by weak(referenceB)
        val element = Subscription(
            listener = { item: T ->
                val a = weakA;
                val b = weakB;
                if (a != null && b != null) {
                    listener(a, b, item);
                    return@Subscription false
                } else {
                    return@Subscription true
                }
            },
            identifier = id
        )
        id += 1
        subscriptions.add(element)
        return element
    }

    fun <A : AnyObject, B : AnyObject, C : AnyObject> addWeak(
        referenceA: A,
        referenceB: B,
        referenceC: C, @escaping listener: (A, B, C, T) -> Unit
    ): Subscription<T> {
        val weakA by weak(referenceA)
        val weakB by weak(referenceB)
        val weakC by weak(referenceC)
        val element = Subscription(
            listener = { item: T ->
                val a = weakA;
                val b = weakB;
                val c = weakC;
                if (a != null && b != null && c != null) {
                    listener(a, b, c, item);
                    return@Subscription false
                } else {
                    return@Subscription true
                }
            },
            identifier = id
        )
        id += 1
        subscriptions.add(element)
        return element
    }


    fun <A : AnyObject> addAndRunWeak(
        referenceA: A,
        value: T,
        @escaping listener: (A, T) -> Unit
    ): Subscription<T> {
        listener(referenceA, value)
        val weakA by weak(referenceA);
        val element = Subscription(
            listener = { item: T ->
                val a = weakA;
                if (a != null) {
                    listener(a, item);
                    return@Subscription false
                } else {
                    return@Subscription true
                }
            },
            identifier = id
        )
        id += 1
        subscriptions.add(element)
        return element
    }

    fun <A : AnyObject, B : AnyObject> addAndRunWeak(
        referenceA: A,
        referenceB: B,
        value: T,
        @escaping listener: (A, B, T) -> Unit
    ): Subscription<T> {
        listener(referenceA, referenceB, value)
        val weakA by weak(referenceA)
        val weakB by weak(referenceB)
        val element = Subscription(
            listener = { item: T ->
                val a = weakA;
                val b = weakB;
                if (a != null && b != null) {
                    listener(a, b, item);
                    return@Subscription false
                } else {
                    return@Subscription true
                }
            },
            identifier = id
        )
        id += 1
        subscriptions.add(element)
        return element
    }

    fun <A : AnyObject, B : AnyObject, C : AnyObject> addAndRunWeak(
        referenceA: A,
        referenceB: B,
        referenceC: C,
        value: T,
        @escaping listener: (A, B, C, T) -> Unit
    ): Subscription<T> {
        listener(referenceA, referenceB, referenceC, value)
        val weakA by weak(referenceA)
        val weakB by weak(referenceB)
        val weakC by weak(referenceC)
        val element = Subscription(
            listener = { item: T ->
                val a = weakA;
                val b = weakB;
                val c = weakC;
                if (a != null && b != null && c != null) {
                    listener(a, b, c, item);
                    return@Subscription false
                } else {
                    return@Subscription true
                }
            },
            identifier = id
        )
        id += 1
        subscriptions.add(element)
        return element
    }
}
